src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -184,7 +184,15 @@ private function parseDefinition(\DOMElement $service, $file)
                 } elseif ($childService = $factory->getAttribute('service')) {
                     $class = new Reference($childService, ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE, false);
                 } else {
-                    $class = $factory->getAttribute('class');
+                    $class = $factory->getAttribute('class') ?: $definition->getClass();
+                }
+
+                if (null === $class) {
```

Comment:
This check is only here to provide more details information (file, service, id, ...).
A similar exception would be thrown from the definition, but not DX friendly because those information are lost.